lib/zz.mli

```diff
@@ -0,0 +1,183 @@
+
+     {[
+let rec go d0 = match M.decod d0 with
+  | `Await d1 -> ... go d1
+  | `Flush d1 -> ... go d1
+  | _ -> ... in
+     ]} *)
+
+module M : sig
+  type decoder
+  (** The type for decoders. *)
+
+  type src = [ `Channel of in_channel | `String of string | `Manual ]
+  (** The type for input sources. With a [`Manual] source the client must
+     provide input with {!src}. With [`String] or [`Channel] source the client
+     can safely discard [`Await] case (with [assert false]). *)
+
+  type decode = [ `Await of decoder | `Flush of decoder | `End of decoder | `Malformed of string ]
+
+  val decoder : src -> o:bigstring -> allocate:(int -> window) -> decoder
+  (** [decoder src ~o ~allocate] is a decoder that inputs from [src].
+
+      {b Output buffer.}
+
+      [zz], as [dd], uses [o] buffer as internal buffer to store output. We
+     recommend to allocate an {!io_buffer_size} buffer as output buffer. Then,
+     {!dst_rem} gives you how many bytes it remains in [o].
+
+      {b Window.}
+
+      ZLIB has an header to notice length of window needed to inflate given
+     input. When [zz] knows that, it calls [allocate] with a number [bits] so
+     that [1 lsl bits] is length of window. [bits] can not be upper than 15 and
+     lower than 8. [allocate] can be [fun bits -> Dd.make_window ~bits] or a
+     previously allocated window. [decoder] will take the ownership on it! *)
+
+  val decode : decoder -> decode
+  (** [decode d0] is:
+
+      {ul
+      {- [`Await d1] if [d0] has a [`Manual] input source and awaits for more
+     input. The client must use {!src} with [d1] to provide it.}
+      {- [`Flush d1] if given output buffer [o] (see {!decoder}) is full. The
+     client must use {!flush} with [d1] to {b completely} flush [o]. {!dst_rem}
+     gives you how many bytes it remains in [o]. [M.dst_rem d1 - bigstring_length o]
+     gives you how many bytes are available.}
+      {- [`Malformed err] if given input is malformed. [err] is a human-readable
+     error.}
+      {- [`End d1] if given input notify end of flow. [o] is possibly not
+     empty (it can be check with {!dst_rem}).}} *)
+
+  val reset : decoder -> decoder
+  (** [reset d] is a decoder as is when it was created by {!decoder}. *)
+
+  val src : decoder -> bigstring -> int -> int -> decoder
+  (** [src d s j l] provides [d] with [l] bytes to read, starting at [j] in [s].
+     This byte range is read by calls to {!decode} with [d] until [`Await] is
+     returned. To signal the end of input call the function with [l = 0].
+
+     @raise Invalid_argument when [j] and [l] do not correspond to a valid
+     range. *)
+
+  val dst_rem : decoder -> int
+  (** [dst_rem d] is how many bytes it remains in given output buffer [o]. *)
+
+  val src_rem : decoder -> int
+  (** [src_rem d] is how many bytes it remains in given input buffer. *)
+
+  val write : decoder -> int
+  (** [write d] is how many bytes [d] emitted since it was created. *)
```

Comment:
The question here is much more complex. Indeed if you look into details (mostly `Dd.N` implementation), the code is currently not really compatible with a 32-bits architecture. @Julow gives me some ideas but the cost on performance should be big.

This situation is not very clear/stable and if you think that the support of 32-bits architecture is obsolete, we can enforce that `decompress` should be available only on 64 bits architecture. But I'm not sure that is the best solution. @mirage/core?

Reply:
I think dropping 32bit support would be sad, especially since people are currently working on e.g. ESP32 compilation backends, and it would be a shame to not be able to use this library there, since quite a lot of things depend on `decompress`.